### PR TITLE
BUG: remove outdated code exposed by a refactor

### LINKED ISSUE #15753: assert_equal fails/crashes on empty arrays with large shape
```py
>>> a = np.array([], dtype=np.int64).reshape((6, 10, 6, 10, 0, 5, 1, 9, 10, 2, 6, 2, 6, 0, 10, 9, 5, 1, 5, 1, 10, 3))
>>> np.testing.assert_equal(a, a[()])
python(73392,0x113ae6dc0) malloc: can't allocate region
*** mach_vm_map(size=15746400002048) failed (error code=3)
python(73392,0x113ae6dc0) malloc: *** set a breakpoint in malloc_error_break to debug
python(73392,0x113ae6dc0) malloc: can't allocate region
*** mach_vm_map(size=15746400002048) failed (error code=3)
python(73392,0x113ae6dc0) malloc: *** set a breakpoint in malloc_error_break to debug
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/aaronmeurer/anaconda3/lib/python3.7/site-packages/numpy/testing/_private/utils.py", line 349, in assert_equal
    return assert_array_equal(actual, desired, err_msg, verbose)
  File "/Users/aaronmeurer/anaconda3/lib/python3.7/site-packages/numpy/testing/_private/utils.py", line 936, in assert_array_equal
    verbose=verbose, header='Arrays are not equal')
  File "/Users/aaronmeurer/anaconda3/lib/python3.7/site-packages/numpy/testing/_private/utils.py", line 770, in assert_array_compare
    flagged = func_assert_same_pos(x, y, func=isnan, hasval='nan')
  File "/Users/aaronmeurer/anaconda3/lib/python3.7/site-packages/numpy/testing/_private/utils.py", line 746, in func_assert_same_pos
    raise AssertionError(msg)
AssertionError:
Arrays are not equal

x and y nan location mismatch:
 x: array([], shape=(6, 10, 6, 10, 0, 5, 1, 9, 10, 2, 6, 2, 6, 0, 10, 9, 5, 1, 5, 1, 10, 3),
      dtype=int64)
 y: array([], shape=(6, 10, 6, 10, 0, 5, 1, 9, 10, 2, 6, 2, 6, 0, 10, 9, 5, 1, 5, 1, 10, 3),
      dtype=int64)
```

I also sometimes get MemoryError from similar examples, for instance, with `shape=[7, 2, 6, 10, 6, 10, 0, 5, 1, 9, 10, 2, 6, 2, 6, 0, 10, 9, 5, 1, 5, 1, 10, 3]`, and even `np.all` gives warnings from malloc.

```py
>>> a = np.array([], dtype=np.int64).reshape([7, 2, 6, 10, 6, 10, 0, 5, 1, 9, 10, 2, 6, 2, 6, 0, 10, 9, 5, 1, 5, 1, 10, 3])
>>> np.all(a == a[()])
python(77502,0x10cb79dc0) malloc: can't allocate region
*** mach_vm_map(size=220449600000000) failed (error code=3)
python(77502,0x10cb79dc0) malloc: *** set a breakpoint in malloc_error_break to debug
python(77502,0x10cb79dc0) malloc: can't allocate region
*** mach_vm_map(size=220449600000000) failed (error code=3)
python(77502,0x10cb79dc0) malloc: *** set a breakpoint in malloc_error_break to debug
False
```

```py
>>> np.testing.assert_equal(a, a[()])
python(77502,0x10cb79dc0) malloc: can't allocate region
*** mach_vm_map(size=220449600000000) failed (error code=3)
python(77502,0x10cb79dc0) malloc: *** set a breakpoint in malloc_error_break to debug
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/aaronmeurer/anaconda3/lib/python3.7/site-packages/numpy/testing/_private/utils.py", line 349, in assert_equal
    return assert_array_equal(actual, desired, err_msg, verbose)
  File "/Users/aaronmeurer/anaconda3/lib/python3.7/site-packages/numpy/testing/_private/utils.py", line 936, in assert_array_equal
    verbose=verbose, header='Arrays are not equal')
  File "/Users/aaronmeurer/anaconda3/lib/python3.7/site-packages/numpy/testing/_private/utils.py", line 770, in assert_array_compare
    flagged = func_assert_same_pos(x, y, func=isnan, hasval='nan')
  File "/Users/aaronmeurer/anaconda3/lib/python3.7/site-packages/numpy/testing/_private/utils.py", line 728, in func_assert_same_pos
    x_id = func(x)
MemoryError: Unable to allocate 0 bytes for an array with shape (7, 2, 6, 10, 6, 10, 0, 5, 1, 9, 10, 2, 6, 2, 6, 0, 10, 9, 5, 1, 5, 1, 10, 3) and data type bool
```

Note:

```py
>>> shape = [7, 2, 6, 10, 6, 10, 0, 5, 1, 9, 10, 2, 6, 2, 6, 0, 10, 9, 5, 1, 5, 1, 10, 3]
>>> np.prod([i for i in shape if i])
220449600000000
```

It isn't clear to me if this is the sort of thing that should work. I would expect empty arrays to effectively ignore the nonzero entries in the shape except for broadcasting purposes. But maybe there are technical reasons why this cannot happen?

### Numpy/Python version information:

```py
>>> import sys, numpy; print(numpy.__version__, sys.version)
1.18.1 3.7.3 | packaged by conda-forge | (default, Mar 27 2019, 15:43:19)
[Clang 4.0.1 (tags/RELEASE_401/final)]
```

### PR DESCRIPTION
**Pull request by mattip**
Closes #15753 

The removed code should have been removed long ago. The zero case was masked because `nbytes` (before gh-7463) was reset to the value returned from `_array_fill_strides`. In gh-7463 `_array_fill_strides` was refactored and no longer returns a value, so `nbytes` was never zeroed out, leading to the bug.

### CODE REVIEW

**Comment by mattip** [comment]
Theoretically this should be backported to anything after gh-7463 from 2016, but since the issue was only opened recently I think it is fine to leave this for 1.19. I think it is "only" a performance hit (memory is allocated then released) for cases with shapes that fit into memory.

File: numpy/core/src/multiarray/ctors.c [hunk 1]
```diff
@@ -985,14 +985,6 @@ PyArray_NewFromDescr_int(
...
         npy_intp dim = dims[i];
 
-        if (dim == 0) {
-            /*
-             * Compare to PyArray_OverflowMultiplyList that
-             * returns 0 in this case.
-             */
-            continue;
-        }
-
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/ctors.c]: To some extent this might be deliberate - it means that `arr[:,i].data` still points within allocated memory, rather than being an invalid pointer.

Another way to ensure this would be to just set all the strides to zero.

Perhaps it doesn't matter.

**Review by eric-wieser — COMMENTED** [review]
The test doesn't appear to fail on master with windows for me.

I suspect this will also produce an unreasonably large allocation:
```
np.array([], dtype=[('a', (np.uint8, (2**31-1)))])
```
but not a large enough one to cause a crash on all systems, since we're limited to 2**32 here.

**Comment by mattip** [comment]
Maybe your malloc only allocates on write? If you can successfully call `npy_alloc_cache(nbytes)` where `nbytes = 1000**4` you must have a lot of memory. What python/os are you using?

**Comment by eric-wieser** [comment]
Nevermind, I missed that the call to `np.isnan` is the one that actually performs the offending allocation.

File: numpy/core/src/multiarray/ctors.c (hunk 1 above)
Comment (mattip) [reply numpy/core/src/multiarray/ctors.c]: nbytes is only used for new allocations, not for views.

Comment (eric-wieser) [reply numpy/core/src/multiarray/ctors.c]: Right, but the point is that when I do `view = arr[:,i]` I end up with `view.data = arr.data + i * arr.strides[1]`. The construction here ensures that `view.data` lies within the chunk of memory allocated by `arr.data`.

Comment (mattip) [reply numpy/core/src/multiarray/ctors.c]: I don't see how this PR changes anything. The loop in question only is for `nbytes`, and `nbytes` is only used for new arrays, not for views. What am I missing?

Comment (eric-wieser) [reply numpy/core/src/multiarray/ctors.c]: Let's imagine that arrays have a new `allocation_size` attribute, which is the number passed to `malloc`.
```
>>> a = np.zeros((2, 3, 0), np.uint8)
>>> a.allocation_size
6  # before this patch
1  # after this patch
>>> a.strides
(3, 1, 1)
>>> b = a[1:]
>>> diff = b.ctypes.data - a.ctypes.data
3  # same before and after this patch
>>>0 <= diff <= a.allocation_size
True  # before this patch
False  # after this patch
```

Unfortunately, if that last condition is false then we invoked undefined behavior at the point we computed `b.ctypes.data`

Comment (mattip) [reply numpy/core/src/multiarray/ctors.c]: `b = a[1:]` does nothing with `nbytes`. The code skips over the allocation stanza since `data != NULL`. I checked in a debugger.

Comment (mattip) [reply numpy/core/src/multiarray/ctors.c]: refactored the code to make the use of `nbytes` clearer

Comment (seberg) [reply numpy/core/src/multiarray/ctors.c]: Yeah, I think we may have actually said that this may be just as well when putting this in, hmmm (I do not remember). It seems a bit of a tiny issue: what is the use-case for huge but empty arrays, when the non-empty version would be too big to handle?
In fact, I think Eric is right, you could have to arrays which may appear to `may_share_memory`, unless we have a special case that empty arrays cannot share memory. Maybe it would be better to use `NULL` as the data pointer for empty arrays if we are going here, I hope that would not kill any users, but at least it would be obvious if there is incorrect use?

Comment (eric-wieser) [reply numpy/core/src/multiarray/ctors.c]: @mattip, the issue is not what `a[1:]` does, but that as a result of your change the allocation within `a` is now dangerously small, and the bits of numpy that handle computing views invoke undefined behavior.

Comment (seberg) [reply numpy/core/src/multiarray/ctors.c]: Sorry, yeah, NULL would require way too many special cases.

Comment (eric-wieser) [reply numpy/core/src/multiarray/ctors.c]: > I think we may have actually said that this may be just as well when putting this in

Note the offending code has been around for a long time, although github is serving me unicorns if I attempt to go back further: https://github.com/numpy/numpy/blob/91be9a9ace3e532a40a93f0c00d71fa34adce009%5E/numpy/core/src/arrayobject.c#L4618

Comment (seberg) [reply numpy/core/src/multiarray/ctors.c]: Hmm, true, so the actual change is just that there are more copies of these unused mallocs around? Also `_array_fill_strides` which used to be use, has the same behaviour since the dawn of time.

Comment (mattip) [reply numpy/core/src/multiarray/ctors.c]: The refactor fails, since we apparently need the error to be raised even for views.

Comment (mattip) [reply numpy/core/src/multiarray/ctors.c]: @eric-wieser could you expand on "the allocation within `a` is now dangerously small" ? As I understand it, either `b` is a view into `a`, in which case its `data` is NULL, or `b` has its own memory. The only case where `b.base` and `b.data` both are non-NULL is for `WRITEBACKIFCOPY`.

Comment (seberg) [reply numpy/core/src/multiarray/ctors.c]: Data is always non-null, but we normally make sure it points to valid memory even for empty arrays. With this change, that will not be true anymore, e.g.:
```
In [1]: arr = np.ones((2**59, 0), dtype=np.uint8)                                                                       
In [2]: arr2 = np.ones(100)                                                                                             
In [3]: diff = arr2.__array_interface__["data"][0] - arr.__array_interface__["data"][0]                                 

In [4]: arr[diff:].__array_interface__["data"][0]                                                                       
Out[4]: 24742928

In [5]: arr2.__array_interface__["data"][0]                                                                             
Out[5]: 24742928
```
`arr[diff:]` has nothing to do with `arr2`, but can be made to point to its memory. Now of course nobody should be accessing `arr[diff:]`'s memory. `np.may_share_memory` seems to survive this, although I wonder if that is due to some fastpath and not an actual general behaviour.

Comment (mattip) [reply numpy/core/src/multiarray/ctors.c]: Ahh, I see, the problem is when allocating `a` not when getting a view into it. The code in master will allocate "as if" the `0` values in the shape were replaced with `1`, so `np.zeros((0, 1000, 1000, 1000, 1000))` throws the error `MemoryError: Unable to allocate 0 bytes for an array with shape (0, 1000, 1000, 1000, 1000) and data type float64`. Now when taking a view into this, `b.ctypes.data` will point to somewhere inside the allocation. Is there a way to actually reference the memory allocated? Any `__getitem__` should fail. Using `ctypes.cast` is cheating, since I could just cast any integer to a pointer using that construct.

Comment (eric-wieser) [reply numpy/core/src/multiarray/ctors.c]: > Is there a way to actually reference the memory allocated?

It's not relevant, because the C standard considers the following undefined behavior even though no dereference occurs:
```C
int x[4] = {1, 2, 3, 4];
int *end = x + 4;  // ok
int *past_end = x + 5;  // undefined behavior here
```

Comment (mattip) [reply numpy/core/src/multiarray/ctors.c]: `np.may_share_memory` has a [fast path](https://github.com/numpy/numpy/blob/v1.18.2/numpy/core/src/common/mem_overlap.c#L668) for `if dims[i] == 0`. Don't we consider places that allow access to memory beyond `strides[i] * dims[i]` to be a bug?

Comment (seberg) [reply numpy/core/src/multiarray/ctors.c]: Ah, I was not aware that the C-standard complains about it. If that is the case, there is probably no point in spending too much time on this? It will probably just work TM, but I honestly do not see the big issue with the current behaviour (or in other words: is there an actual use-case?)

Comment (eric-wieser) [reply numpy/core/src/multiarray/ctors.c]: @seberg: "current" meaning master or this patch?

Comment (seberg) [reply numpy/core/src/multiarray/ctors.c]: Master behaviour

Comment (mattip) [reply numpy/core/src/multiarray/ctors.c]: The fact that `a.ctypes.data != b.ctypes.data` in your example above seems like another bug. If `a.size == 0` then I would expect any view of `a` to point to the same, if not NULL, `data`

Comment (eric-wieser) [reply numpy/core/src/multiarray/ctors.c]: I'm not sure I'd consider it a bug, but option 2 from https://github.com/numpy/numpy/pull/15780#discussion_r395122076 would make the pointers equal.

Comment (mattip) [reply numpy/core/src/multiarray/ctors.c]: fixed so `a.ctypes.data == b.ctypes.data` if `any(a.shape ==0)` in 39e1233

File: numpy/core/src/multiarray/mapping.c [hunk 2]
```diff
@@ -912,38 +912,53 @@ get_view_from_index(PyArrayObject *self, PyArrayObject **view,
...
+     * prevent ptr + offset going beyond the end of the memory on empty arrays,
+     * which would be undefined behaviour according to the C99 spec paragraph
+     * 6.5.6 http://www.open-std.org/jtc1/sc22/wg14/www/docs/n1256.pdf
+     */
+    for (i=0; i < PyArray_NDIM(self); i++) {
+        if (self_dims[i] == 0) {
+            is_empty = 1;
+        }
+    }
+
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/mapping.c]: Wouldn't it be easier just to construct the strides on an empty array so that this doesn't happen?

Comment (eric-wieser) [reply numpy/core/src/multiarray/mapping.c]: Also safer, since we probably perform this type of addition all over the place.

Comment (mattip) [reply numpy/core/src/multiarray/mapping.c]: strides seem to be mutable with little restrictions on their values. Dimensions less so.
```
a = np.zeros((1, 1, 0))
a.strides = (10000, 32, 12)  # passes
a.shape = (1, 1, 1)  #errors
```

Comment (eric-wieser) [reply numpy/core/src/multiarray/mapping.c]: If strides are really that mutable, then the users can already invoke undefined behavior anyway.

Comment (mattip) [reply numpy/core/src/multiarray/mapping.c]: (This only applies to `a.size == 0` strides, otherwise there are checks in place to make sure `a.strides` does not allow access outside `a.data`). I am less concerned about undefined behaviour. I think 39e1233  is an improvement since it doesn't make sense for a view of a 0-sized array to point to different memory. I think this is (1) from the comment above

Comment (mattip) [reply numpy/core/src/multiarray/mapping.c]: Other places we use `PyArray_NewFromDescr{,AndBase}` with a possibly modified data pointer, all others call with `NULL` or with a `PyArray_DATA()` result: 
- `VOID_getitem`, `VOID_setitem` (`ip + offset` for structured dtypes),
- `arr_unravel_index` (seems safe),
- `PyArray_FromInterface` (the use of `offset` here looks suspicious, there are no checks that it is valid),
- `PyArray_FromBuffer` (checks validity), 
- `_get_part` (the implementation of `a.real` or `a.imag`, calculated internally and seems valid)
- `PyArray_Diagonal` (calculated internally, seems valid)
- `PyArray_Nonzero` (safe, the offset will always be 0 since `nonzero` is 0)
- `iter_array`, `npyiter_seq_item`, `npyiter_seq_ass_item` and `NpyIter_GetIterView` (too complicated to to dive into for this comment)
- `ufunc_at` (too complidated to dive into for this comment)

Additionally, the following API functions accept a data pointer `PArray_New`, `PyArray_CastScalarToCtypes`, anything goes with these.

So in conclusion, the only known dangerous one seems to be `offset` in `__array_interface__`, which we should probably have deprecated instead of fixed in gh-13036.

File: numpy/core/src/multiarray/ctors.c (hunk 1 above)
Comment (seberg) [reply numpy/core/src/multiarray/ctors.c]: OK, I am not against this. One thing though: Some operation (such as reductions along the empty axis) want to combine the other axes (for iteration). The current check ensures that the total number of elements is limited by the size of `intp`. Removing this will remove that check, which means that such "reshapes" can fail. Maybe we can just ignore that, but certain operations will fail on such too-large arrays.

**Comment by mattip** [comment]
Closing, this is not close enough to a correct fix
